smt_parser.py

```diff
@@ -0,0 +1,264 @@
+"""
+This module is a smt2 parser built upon a parser combinator library present in parser_combinator.py. 
+
+
+In order to keep the parser nice and readable we establish a few rules used for development outlined below:
+    1. whitespace parsing pushed down the parsing stack as far as possible
+    2. avoid writing complicated parsers functions, prefer the functional pmap for short parsers and prefer avoiding pmap for longer parsers
+        for a pmap style parser see parse_type_builtin_bool and for the other style see parse_type_bit_vec
+    3. if an operator is commonly used, feel free to add a shorthand to keep things more readable (example ws)
+"""
+import parser_combinator as pc
+import smt
+import source
+import assume_prove
+import re
+import typing as tp
+
+
+# shortand for parser that munches whitespace since we use this call a lot
+ws = pc.without_ws
+
+
+def parse_op() -> pc.Parser[source.Operator]:
+    ps = []
+
+    def get_ret_fn(op: source.Operator) -> tp.Callable[[str], source.Operator]:
+        def fn(_: str) -> source.Operator:
+            return op
+        return fn
+
+    for op, value in smt.ops_to_smt.items():
+        strValue = str(value)
+        fn = get_ret_fn(op)
+        p = pc.pmap(ws(pc.string(strValue)), fn)
+        ps.append(p)
+    return pc.choice(ps)
+
+
+def parse_integer() -> pc.Parser[int]:
+    def to_int(s: str) -> int:
+        return int(s, 10)
+    return pc.pmap(ws(pc.regex(re.compile(r"[0-9]+"))), to_int)
```

Comment:
Only because in the smt_parser I just used the regex provided in the smt lib for integer parsing, this is also the case for the hex nums, pc does have a pc.integer(). I just changed it to that now